plugins/metalsmith-revision.js

```diff
@@ -1,85 +1,53 @@
-const resolve = require('path').resolve;
-const relative = require('path').relative;
-const fs = require('fs');
-const crypto = require('crypto');
-const debug = require('debug')('metalsmith-revision');
-
-const defaultOptions = { layout: false, layoutDir: './layouts' };
-
-let initialized = false;
-
-function plugin(opts) {
-  const options = Object.assign({}, defaultOptions, opts);
-  return (files, metalsmith, done) => {
-
-    const configPath = resolve(metalsmith._directory, '.revision');
-    const hashTable = { layouts: {}, src: {}, frontmatter: {} };
-    const layoutDirectory = resolve(metalsmith._directory, options.layoutDir);
-
-    if(!initialized && fs.existsSync(configPath)) {
-      initialized = true;
-      fs.unlinkSync(configPath)
-      debug('Refreshing revision file');
+const { resolve } = require("path");
+const fs = require("fs");
+const crypto = require("crypto");
+
+const checksum = (str) =>
+  crypto.createHash("md5").update(str, "utf8").digest("hex");
+
+// attrubutes that are not used to generate a cache-hash.
+const attrsToIgnore = ["stats", "model", "filename"];
+
+module.exports = function (files, metalsmith, done) {
```

Comment:
### TL;DR

* if any `meta` a.k.a "frontmatter" changed, we rebuild everything
* if some `contents` changed, we rebuild only those pages

----------------

longer story: we place a `.revision` file in the root folder (as that is what the previous version did). that file contains two dicts - `contents` and `meta`. `contents` maps each file to a checksum of its contents, `meta` maps each file to a checksum of its frontmatter-entries. this is necessary to e.g. invalidate this cache in case you change the `title` or `navigationTitle` or whatnot, as such change needs to propagate to every page (cause of the way out layouts work).

as some `meta` entries are not something we specify, we strip them via `attrsToIgnore`.

-------------

groking this is kinda hard i guess, so feel free to reach out via zoom and i walk you through, if you really want to understand what's going on here.